File: streampark-flink/streampark-flink-packer/src/main/scala/org/apache/streampark/flink/packer/pipeline/impl/FlinkK8sApplicationBuildPipeline.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.flink.packer.pipeline.impl

import org.apache.streampark.common.fs.LfsOperator
import org.apache.streampark.common.util.ThreadUtils
import org.apache.streampark.flink.kubernetes.PodTemplateTool
import org.apache.streampark.flink.kubernetes.ingress.IngressController
import org.apache.streampark.flink.packer.docker._
import org.apache.streampark.flink.packer.maven.MavenTool
import org.apache.streampark.flink.packer.pipeline._
import org.apache.streampark.flink.packer.pipeline.BuildPipeline.executor

import com.github.dockerjava.api.command.PushImageCmd
import com.github.dockerjava.core.command.{HackBuildImageCmd, HackPullImageCmd, HackPushImageCmd}
import com.google.common.collect.Sets
import org.apache.commons.lang3.StringUtils

import java.io.File
import java.util.concurrent.{LinkedBlockingQueue, ThreadPoolExecutor, TimeUnit}

import scala.concurrent.{ExecutionContext, Future}

/** Building pipeline for flink kubernetes-native application mode */
class FlinkK8sApplicationBuildPipeline(request: FlinkK8sApplicationBuildRequest)
  extends BuildPipeline {

  override def pipeType: PipelineTypeEnum =
    PipelineTypeEnum.FLINK_NATIVE_K8S_APPLICATION

  private var dockerProcessWatcher: DockerProgressWatcher =
    new SilentDockerProgressWatcher

  // non-thread-safe
  private val dockerProcess = new DockerResolveProgress(
    DockerPullProgress.empty(),
    DockerBuildProgress.empty(),
    DockerPushProgress.empty())

  override protected def offerBuildParam: FlinkK8sApplicationBuildRequest =
    request

  def registerDockerProgressWatcher(watcher: DockerProgressWatcher): Unit = {
    dockerProcessWatcher = watcher
  }

  @throws[Throwable]
  override protected def buildProcess(): DockerImageBuildResponse = {

    // Step-1: init build workspace of flink job
    // the sub workspace dir like: APP_WORKSPACE/k8s-clusterId@k8s-namespace/
    val buildWorkspace =
      execStep(1) {
        val buildWorkspace =
          s"${request.workspace}/${request.clusterId}@${request.k8sNamespace}"
        LfsOperator.mkCleanDirs(buildWorkspace)
        logInfo(s"Recreate building workspace: $buildWorkspace")
        buildWorkspace
      }.getOrElse(throw getError.exception)

    // Step-2: export k8s pod template files
    val podTemplatePaths = request.flinkPodTemplate match {
      case podTemplate if podTemplate.isEmpty =>
        skipStep(2)
        Map[String, String]()
      case podTemplate =>
        execStep(2) {
          val podTemplateFiles =
            PodTemplateTool
              .preparePodTemplateFiles(buildWorkspace, podTemplate)
              .tmplFiles
          logInfo(s"Export flink podTemplates: ${podTemplateFiles.values.mkString(",")}")
          podTemplateFiles
        }.getOrElse(throw getError.exception)
    }

    // Step-3: build shaded flink job jar and handle extra jars
    // the output shaded jar file name like: streampark-flinkjob_myjob-test.jar
    val (shadedJar, extJarLibs) =
      execStep(3) {
        val shadedJarOutputPath = request.getShadedJarPath(buildWorkspace)
        val shadedJar =
          MavenTool.buildFatJar(request.mainClass, request.providedLibs, shadedJarOutputPath)
        logInfo(s"Output shaded flink job jar: ${shadedJar.getAbsolutePath}")
        shadedJar -> request.dependencyInfo.extJarLibs
      }.getOrElse(throw getError.exception)

    // Step-4: generate and Export flink image dockerfiles
    val (dockerfile, dockerFileTemplate) =
      execStep(4) {
        val dockerFileTemplate = {
          if (request.integrateWithHadoop) {
            FlinkHadoopDockerfileTemplate.fromSystemHadoopConf(
              buildWorkspace,
              request.flinkBaseImage,
              shadedJar.getAbsolutePath,
              extJarLibs)
          } else {
            FlinkDockerfileTemplate(
              buildWorkspace,
              request.flinkBaseImage,
              shadedJar.getAbsolutePath,
              extJarLibs)
          }
        }
        val dockerFile = dockerFileTemplate.writeDockerfile
        logInfo(
          s"Output flink dockerfile: ${dockerFile.getAbsolutePath}, content: \n${dockerFileTemplate.offerDockerfileContent}")
        dockerFile -> dockerFileTemplate
      }.getOrElse(throw getError.exception)

    val dockerConf = request.dockerConfig
    val baseImageTag = request.flinkBaseImage.trim
    val pushImageTag = {
      if (request.k8sNamespace.isEmpty || request.clusterId.isEmpty) {
        throw new IllegalArgumentException("k8sNamespace or clusterId cannot be empty")
      }
      val expectedImageTag =
        s"streampark-flinkjob-${request.k8sNamespace}-${request.clusterId}"
      compileTag(expectedImageTag, dockerConf.registerAddress, dockerConf.imageNamespace)
    }

    // Step-5: pull flink base image
    execStep(5) {
      usingDockerClient {
        dockerClient =>
          val pullImageCmd = {
            // when the register address prefix is explicitly identified on base image tag,
            // the user's pre-saved docker register auth info would be used.
            val pullImageCmdState =
              dockerConf.registerAddress != null && !baseImageTag.startsWith(
                dockerConf.registerAddress)
            if (pullImageCmdState) {
              dockerClient.pullImageCmd(baseImageTag)
            } else {
              dockerClient
                .pullImageCmd(baseImageTag)
                .withAuthConfig(dockerConf.toAuthConf)
            }
          }
          val pullCmdCallback = pullImageCmd
            .asInstanceOf[HackPullImageCmd]
            .start(watchDockerPullProcess {
              pullRsp =>
                dockerProcess.pull.update(pullRsp)
                Future(dockerProcessWatcher.onDockerPullProgressChange(dockerProcess.pull.snapshot))
            })
          pullCmdCallback.awaitCompletion
          logInfo(s"Already pulled docker image from remote register, imageTag=$baseImageTag")
      }(err => throw new Exception(s"Pull docker image failed, imageTag=$baseImageTag", err))
    }.getOrElse(throw getError.exception)

    // Step-6: build flink image
    execStep(6) {
      usingDockerClient {
        dockerClient =>
          val buildImageCmd = dockerClient
            .buildImageCmd()
            .withBaseDirectory(new File(buildWorkspace))
            .withDockerfile(dockerfile)
            .withTags(Sets.newHashSet(pushImageTag))

          val buildCmdCallback = buildImageCmd
            .asInstanceOf[HackBuildImageCmd]
            .start(watchDockerBuildStep {
              buildStep =>
                dockerProcess.build.update(buildStep)
                Future(
                  dockerProcessWatcher.onDockerBuildProgressChange(dockerProcess.build.snapshot))
            })
          val imageId = buildCmdCallback.awaitImageId
          logInfo(s"Built docker image, imageId=$imageId, imageTag=$pushImageTag")
      }(err => throw new Exception(s"Build docker image failed. tag=$pushImageTag", err))
    }.getOrElse(throw getError.exception)

    // Step-7: push flink image
    execStep(7) {
      usingDockerClient {
        dockerClient =>
          val pushCmd: PushImageCmd = dockerClient
            .pushImageCmd(pushImageTag)
            .withAuthConfig(dockerConf.toAuthConf)

          val pushCmdCallback = pushCmd
            .asInstanceOf[HackPushImageCmd]
            .start(watchDockerPushProcess {
              pushRsp =>
                dockerProcess.push.update(pushRsp)
                Future(dockerProcessWatcher.onDockerPushProgressChange(dockerProcess.push.snapshot))
            })
          pushCmdCallback.awaitCompletion
          logInfo(s"Already pushed docker image, imageTag=$pushImageTag")
      }(err => throw new Exception(s"Push docker image failed. tag=$pushImageTag", err))
    }.getOrElse(throw getError.exception)

    // Step-8:  init build workspace of ingress
    val ingressOutputPath = request.ingressTemplate match {
      case ingress if StringUtils.isBlank(ingress) =>
        skipStep(8)
        ""
      case _ =>
        execStep(8) {
          val ingressOutputPath =
            IngressController.prepareIngressTemplateFiles(buildWorkspace, request.ingressTemplate)
          logInfo(s"Export flink ingress: $ingressOutputPath")
          ingressOutputPath
        }.getOrElse(throw getError.exception)
    }

    DockerImageBuildResponse(
      buildWorkspace,
      pushImageTag,
      podTemplatePaths,
      dockerFileTemplate.innerMainJarPath)
  }

  /** compile image tag with namespace and remote address. */
  private[this] def compileTag(
      tag: String,
      registerAddress: String,
      imageNamespace: String): String = {
    var tagName = if (tag.contains("/")) tag else s"$imageNamespace/$tag"
    val addRegisterAddressState =
      StringUtils.isNotBlank(registerAddress) && !tagName.startsWith(registerAddress)
    if (addRegisterAddressState) {
      tagName = s"$registerAddress/$tagName"
    }
    tagName.toLowerCase
  }

}

object FlinkK8sApplicationBuildPipeline {

  val execPool = new ThreadPoolExecutor(
    Runtime.getRuntime.availableProcessors * 5,
    Runtime.getRuntime.availableProcessors() * 10,
    60L,
    TimeUnit.SECONDS,
    new LinkedBlockingQueue[Runnable](2048),
    ThreadUtils.threadFactory("streampark-docker-progress-watcher-executor"),
    new ThreadPoolExecutor.DiscardOldestPolicy)

  implicit val executor: ExecutionContext =
    ExecutionContext.fromExecutorService(execPool)

  def of(request: FlinkK8sApplicationBuildRequest): FlinkK8sApplicationBuildPipeline =
    new FlinkK8sApplicationBuildPipeline(request)

}

File: streampark-flink/streampark-flink-packer/src/main/java/org/apache/streampark/flink/packer/pipeline/impl/FlinkK8sApplicationBuildPipeline.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.packer.pipeline.impl;

import org.apache.streampark.common.fs.LfsOperator;
import org.apache.streampark.common.util.ThreadUtils;
import org.apache.streampark.flink.kubernetes.PodTemplateTool;
import org.apache.streampark.flink.kubernetes.ingress.IngressController;
import org.apache.streampark.flink.kubernetes.model.K8sPodTemplates;
import org.apache.streampark.flink.packer.docker.DockerConf;
import org.apache.streampark.flink.packer.docker.DockerUtils;
import org.apache.streampark.flink.packer.docker.FlinkDockerfileTemplate;
import org.apache.streampark.flink.packer.docker.FlinkDockerfileTemplateTrait;
import org.apache.streampark.flink.packer.docker.FlinkHadoopDockerfileTemplate;
import org.apache.streampark.flink.packer.maven.MavenTool;
import org.apache.streampark.flink.packer.pipeline.BuildPipeline;
import org.apache.streampark.flink.packer.pipeline.DockerBuildProgress;
import org.apache.streampark.flink.packer.pipeline.DockerImageBuildResponse;
import org.apache.streampark.flink.packer.pipeline.DockerProgressWatcher;
import org.apache.streampark.flink.packer.pipeline.DockerPullProgress;
import org.apache.streampark.flink.packer.pipeline.DockerPushProgress;
import org.apache.streampark.flink.packer.pipeline.DockerResolveProgress;
import org.apache.streampark.flink.packer.pipeline.FlinkK8sApplicationBuildRequest;
import org.apache.streampark.flink.packer.pipeline.PipelineTypeEnum;
import org.apache.streampark.flink.packer.pipeline.SilentDockerProgressWatcher;

import org.apache.commons.lang3.StringUtils;

import com.github.dockerjava.api.command.PushImageCmd;
import com.github.dockerjava.core.command.HackBuildImageCmd;
import com.github.dockerjava.core.command.HackPullImageCmd;
import com.github.dockerjava.core.command.HackPushImageCmd;
import com.google.common.collect.Sets;

import java.io.File;
import java.util.Collections;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/** Building pipeline for flink kubernetes-native application mode */
public class FlinkK8sApplicationBuildPipeline extends BuildPipeline {

    private static final ExecutorService DOCKER_EXECUTOR =
        new ThreadPoolExecutor(
            Runtime.getRuntime().availableProcessors() * 5,
            Runtime.getRuntime().availableProcessors() * 10,
            60L,
            TimeUnit.SECONDS,
            new LinkedBlockingQueue<>(2048),
            ThreadUtils.threadFactory("streampark-docker-progress-watcher-executor"),
            new ThreadPoolExecutor.DiscardOldestPolicy());

    private final FlinkK8sApplicationBuildRequest request;

    private DockerProgressWatcher dockerProcessWatcher = new SilentDockerProgressWatcher();

    private final DockerResolveProgress dockerProcess =
        new DockerResolveProgress(
            DockerPullProgress.empty(),
            DockerBuildProgress.empty(),
            DockerPushProgress.empty());

    public FlinkK8sApplicationBuildPipeline(FlinkK8sApplicationBuildRequest request) {
        this.request = request;
    }

    @Override
    public PipelineTypeEnum pipeType() {
        return PipelineTypeEnum.FLINK_NATIVE_K8S_APPLICATION;
    }

    @Override
    public FlinkK8sApplicationBuildRequest offerBuildParam() {
        return request;
    }

    public void registerDockerProgressWatcher(DockerProgressWatcher watcher) {
        this.dockerProcessWatcher = watcher;
    }

    @Override
    public DockerImageBuildResponse buildProcess() {
        String buildWorkspace =
            execStep(
                1,
                () -> {
                    String workspace =
                        request.workspace()
                            + "/"
                            + request.clusterId()
                            + "@"
                            + request.k8sNamespace();
                    LfsOperator.mkCleanDirs(workspace);
                    logInfo("Recreate building workspace: " + workspace);
                    return workspace;
                })
                    .orElseThrow(() -> {
                        throw pipelineException();
                    });

        Map<String, String> podTemplatePaths;
        K8sPodTemplates podTemplate = request.flinkPodTemplate();
        if (podTemplate.isEmpty()) {
            skipStep(2);
            podTemplatePaths = Collections.emptyMap();
        } else {
            podTemplatePaths =
                execStep(
                    2,
                    () -> {
                        Map<String, String> podTemplateFiles =
                            PodTemplateTool.preparePodTemplateFiles(buildWorkspace, podTemplate)
                                .tmplFiles();
                        logInfo(
                            "Export flink podTemplates: "
                                + String.join(",", podTemplateFiles.values()));
                        return podTemplateFiles;
                    })
                        .orElseThrow(() -> {
                            throw pipelineException();
                        });
        }

        final File shadedJar =
            execStep(
                3,
                () -> {
                    String shadedJarOutputPath = request.getShadedJarPath(buildWorkspace);
                    File jar =
                        MavenTool.buildFatJar(
                            request.mainClass(),
                            request.providedLibs(),
                            shadedJarOutputPath);
                    logInfo("Output shaded flink job jar: " + jar.getAbsolutePath());
                    return jar;
                })
                    .orElseThrow(
                        () -> {
                            throw pipelineException();
                        });
        final Set<String> extJarLibs = request.dependencyInfo().extJarLibs();

        File dockerfile;
        FlinkDockerfileTemplateTrait dockerFileTemplate;
        Object[] dockerResult =
            execStep(
                4,
                () -> {
                    FlinkDockerfileTemplateTrait template;
                    if (request.integrateWithHadoop()) {
                        template =
                            FlinkHadoopDockerfileTemplate.fromSystemHadoopConf(
                                buildWorkspace,
                                request.flinkBaseImage(),
                                shadedJar.getAbsolutePath(),
                                extJarLibs);
                    } else {
                        template =
                            new FlinkDockerfileTemplate(
                                buildWorkspace,
                                request.flinkBaseImage(),
                                shadedJar.getAbsolutePath(),
                                extJarLibs);
                    }
                    File dockerFile = template.writeDockerfile();
                    logInfo(
                        "Output flink dockerfile: "
                            + dockerFile.getAbsolutePath()
                            + ", content: \n"
                            + template.offerDockerfileContent());
                    return new Object[]{dockerFile, template};
                })
                    .orElseThrow(() -> {
                        throw pipelineException();
                    });
        dockerfile = (File) dockerResult[0];
        dockerFileTemplate = (FlinkDockerfileTemplateTrait) dockerResult[1];

        DockerConf dockerConf = request.dockerConfig();
        String baseImageTag = request.flinkBaseImage().trim();
        if (request.k8sNamespace().isEmpty() || request.clusterId().isEmpty()) {
            throw new IllegalArgumentException("k8sNamespace or clusterId cannot be empty");
        }
        String expectedImageTag =
            "streampark-flinkjob-" + request.k8sNamespace() + "-" + request.clusterId();
        String pushImageTag =
            compileTag(
                expectedImageTag, dockerConf.registerAddress(), dockerConf.imageNamespace());

        execStep(
            5,
            () -> {
                DockerUtils.usingDockerClient(
                    dockerClient -> {
                        boolean imgExists =
                            dockerClient.listImagesCmd().exec().stream()
                                .anyMatch(
                                    image -> image.getRepoTags() != null
                                        && java.util.Arrays.stream(image.getRepoTags())
                                            .anyMatch(tag -> tag.contains(baseImageTag)));
                        if (imgExists) {
                            logInfo(
                                "found local docker image "
                                    + baseImageTag
                                    + ", no need to pull from remote.");
                        } else {
                            HackPullImageCmd pullImageCmd;
                            if (dockerConf.registerAddress() != null
                                && !baseImageTag.startsWith(dockerConf.registerAddress())) {
                                pullImageCmd =
                                    (HackPullImageCmd) dockerClient.pullImageCmd(baseImageTag);
                            } else {
                                pullImageCmd =
                                    (HackPullImageCmd) dockerClient
                                        .pullImageCmd(baseImageTag)
                                        .withAuthConfig(dockerConf.toAuthConf());
                            }
                            try {
                                pullImageCmd
                                    .start(
                                        DockerUtils.watchDockerPullProcess(
                                            pullRsp -> {
                                                dockerProcess.getPull().update(pullRsp);
                                                DOCKER_EXECUTOR.submit(
                                                    () -> dockerProcessWatcher
                                                        .onDockerPullProgressChange(
                                                            dockerProcess.getPull().snapshot()));
                                            }))
                                    .awaitCompletion();
                            } catch (InterruptedException e) {
                                Thread.currentThread().interrupt();
                                throw new RuntimeException(e);
                            }
                            logInfo(
                                "Already pulled docker image from remote register, imageTag="
                                    + baseImageTag);
                        }
                        return null;
                    },
                    err -> {
                        throw new RuntimeException(
                            "Pull docker image failed, imageTag=" + baseImageTag, err);
                    });
                return null;
            })
                .orElseThrow(() -> {
                    throw pipelineException();
                });

        execStep(
            6,
            () -> {
                DockerUtils.usingDockerClient(
                    dockerClient -> {
                        HackBuildImageCmd buildImageCmd =
                            (HackBuildImageCmd) dockerClient
                                .buildImageCmd()
                                .withBaseDirectory(new File(buildWorkspace))
                                .withDockerfile(dockerfile)
                                .withTags(Sets.newHashSet(pushImageTag));
                        String imageId =
                            buildImageCmd
                                .start(
                                    DockerUtils.watchDockerBuildStep(
                                        buildStep -> {
                                            dockerProcess.getBuild().update(buildStep);
                                            DOCKER_EXECUTOR.submit(
                                                () -> dockerProcessWatcher
                                                    .onDockerBuildProgressChange(
                                                        dockerProcess.getBuild().snapshot()));
                                        }))
                                .awaitImageId();
                        logInfo(
                            "Built docker image, imageId="
                                + imageId
                                + ", imageTag="
                                + pushImageTag);
                        return null;
                    },
                    err -> {
                        throw new RuntimeException(
                            "Build docker image failed. tag=" + pushImageTag, err);
                    });
                return null;
            })
                .orElseThrow(() -> {
                    throw pipelineException();
                });

        execStep(
            7,
            () -> {
                DockerUtils.usingDockerClient(
                    dockerClient -> {
                        PushImageCmd pushCmd =
                            dockerClient
                                .pushImageCmd(pushImageTag)
                                .withAuthConfig(dockerConf.toAuthConf());
                        try {
                            ((HackPushImageCmd) pushCmd)
                                .start(
                                    DockerUtils.watchDockerPushProcess(
                                        pushRsp -> {
                                            dockerProcess.getPush().update(pushRsp);
                                            DOCKER_EXECUTOR.submit(
                                                () -> dockerProcessWatcher.onDockerPushProgressChange(
                                                    dockerProcess.getPush().snapshot()));
                                        }))
                                .awaitCompletion();
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                            throw new RuntimeException(e);
                        }
                        logInfo("Already pushed docker image, imageTag=" + pushImageTag);
                        return null;
                    },
                    err -> {
                        throw new RuntimeException(
                            "Push docker image failed. tag=" + pushImageTag, err);
                    });
                return null;
            })
                .orElseThrow(() -> {
                    throw pipelineException();
                });

        if (StringUtils.isBlank(request.ingressTemplate())) {
            skipStep(8);
        } else {
            execStep(
                8,
                () -> {
                    String ingressOutputPath =
                        IngressController.prepareIngressTemplateFiles(
                            buildWorkspace, request.ingressTemplate());
                    logInfo("Export flink ingress: " + ingressOutputPath);
                    return ingressOutputPath;
                })
                    .orElseThrow(() -> {
                        throw pipelineException();
                    });
        }

        return new DockerImageBuildResponse(
            buildWorkspace,
            pushImageTag,
            podTemplatePaths,
            dockerFileTemplate.innerMainJarPath());
    }

    private String compileTag(String tag, String registerAddress, String imageNamespace) {
        String tagName = tag.contains("/") ? tag : imageNamespace + "/" + tag;
        if (StringUtils.isNotBlank(registerAddress) && !tagName.startsWith(registerAddress)) {
            tagName = registerAddress + "/" + tagName;
        }
        return tagName.toLowerCase();
    }

    public static FlinkK8sApplicationBuildPipeline of(FlinkK8sApplicationBuildRequest request) {
        return new FlinkK8sApplicationBuildPipeline(request);
    }
}
